[VAULT] Vault Kubernetes KMS docs (#1997)
hashicorp/web-unified-docs: content/vault/v1.21.x/content/docs/deploy/kubernetes/kms/security.mdx

To communicate with the Kubernetes API server, Vault Kubernetes KMS creates a
Unix socket at a configurable path (default: `/tmp/vault-kube-kms.socket`) with
read/write/execute permissions for the owner and group (`0770`). The default
permissions protect the Kubernets plugin by restricting access to the process
itself (the owner) and other processes running in the same group.  

## Process flow

The Vault Kubernetes KMS process runs on the same Kubernetes control plane node
as the Kubernetes API server and listens on a Unix domain socket for gRPC
requests. Process-to-process communication uses kernel-enforced access controls
based on filesystem permissions.

@include 'ld-images/k8s-kms/plugin-flow.mdx'

Before startup:

1. The Vault administrator uses the Transit plugin to create the KEK as a named
   encryption key in Vault and limits access with a Vault policy.

1. The Vault administrator ties authentication (AppRole) to the policy and
   provides the associated appRole ID and secret ID to the Kubernetes
   administrator.
   - **The role ID** is a non-sensitive identifier that you can configure with a
     flag or environment variable.

>    - **The secret ID** is a sensitive credential stored on the Kubernetes API

Reviewer: The secret ID isn't stored on the Kubernetes API server; it's stored on the control plane node. Just replace that and you should be good.

>      server and read by Vault Kubernetes KMS from a filesystem path.

Reviewer: ```suggestion
   - **The secret ID** is a sensitive credential stored on the Kubernetes control plane node and read by Vault Kubernetes KMS from a filesystem path.
```

1. The Kubernetes administrator uses the Vault server address, appRole ID, and
   secret ID to configure a Vault client that will authenticate connection to
   Vault Kubernetes KMS, automatically renew authentication tokens, and handle

>    re-authentication if the token expires.

Reviewer: ```suggestion
   secret ID to configure Vault Kubernetes KMS.
```

The authentication/renewal-related stuff is covered already below. And the user isn't configuring the Vault client directly; the plugin is.

1. The Kubernetes administrator configures a Unix socket for gRPC communication
   between the Kubernetes API server and Vault Kubernetes KMS.

On startup:

> 1. The Kubernetes API server creates and caches the Vault client.

Reviewer: Let's remove this step as it's not accurate (Vault Kubernetes KMS does this, not the Kubernetes API server) and is already covered by the Vault client-related mention above and below.

1. The Kubernetes API server generates a single DEK seed and uses the client to
   request encryption from Vault Kubernetes KMS.

1. Vault Kubernetes KMS creates and caches a Vault client that authenticates to
   Vault, automatically renew authentication tokens, and handles
   re-authentication if the token expires.

1. The Vault client connects to the Vault server over TLS and authenticates
   using the AppRole authentication details provided by the Kubernetes API

>    server.

Reviewer: ```suggestion
   using the AppRole authentication details provided by the Kubernetes administrator.
```

1. Once authenticated, Vault Kubernetes KMS sends the seed to the Transit plugin
   for encryption. **The KEK never leaves Vault and all cryptographic operations
   involving the KEK happen within the Transit plugin**.

1. The Transit plugin returns the encrypted seed to Vault Kubernetes KMS.

> 1. Vault Kubernetes KMS Reports forwards the encrypted seed to the Kubernetes

Reviewer: ```suggestion
1. Vault Kubernetes KMS reports the encrypted seed to the Kubernetes
```
